AWS RDS IAM authentication with Spring Boot (#706)
ordina-jworks/ordina-jworks.github.io: _posts/2022-06-13-aws-rds-iam-authentication-spring-boot.md

---
layout: post
authors: [yolan_vloeberghs]
title: 'RDS IAM Authentication with Spring Boot - secure password-less database authentication on AWS'
image: /img/2022-06-13-aws-rds-iam-authentication-spring-boot.jpeg
tags: [Spring Boot, AWS, Spring, Security]
category: Cloud
comments: true
---

- [Introduction](#introduction)
- [Database role grant](#database-role-grant)
- [AWS](#aws)
- [Spring Boot](#spring-boot)
- [Conclusion](#conclusion)

## Introduction
Database authentication in Spring Boot has traditionally always been very simple to implement and configure to your needs.

> Whether it is with a custom data source, or just using the `spring.datasource` properties, it was always pretty straight out-of-the-box to get a working database connection.

Reviewer: ```suggestion
Whether it is with a custom data source, or just using the `spring.datasource` properties, it was always pretty straight forward to get a working database connection.
```

Reviewer 2: `straightforward`, single word. :)

Handling the sensitive data such as database passwords in different environments can be tricky, but nowadays, there are enough solutions to tackle this problem in a secure and discrete way.

> However, if you are deploying your application on the AWS cloud, you don't have to struggle with your password anymore or need a fancy tool, because you simply don't need a password.

Reviewer: ```suggestion
However, if you are deploying your application on the AWS cloud, you don't have to struggle passwords anymore or need a fancy tool, because you simply don't need a password.
```

> You are now able to establish a database connection by authenticating through IAM. Note that this feature only works for MariaDB, MySQL and PostgreSQL. 

Reviewer: ```suggestion
You are now able to establish a database connection by authenticating through IAM. 
Note that this feature only works for MariaDB, MySQL and PostgreSQL. 
```

    }

    private URI parseJdbcURL(String jdbcUrl) {
        String uri = jdbcUrl.substring(5);
        return URI.create(uri);
    }
}
```

To make sure that our connection does not drop due to authentication failures, we are going to set the maximum lifespan for the connection (remember that authentication tokens are only valid for 15 minutes).

Don't forget to set your database properties (host, port and username). 
**Important note:** IAM authentication to the database requires an SSL connection.

```properties
# Sets the max lifetime of each connection to 840000 ms (14 minutes). After 14 minutes, the application does a new request to RDS for a fresh authentication token
spring.datasource.hikari.max-lifetime=840000
# JDBC url with the properties 'ssl' and 'sslmode', this properties are bound to PostgreSQL instances
spring.datasource.url=jdbc:postgresql://<db-instance-name>.<random>.<region>.rds.amazonaws.com:<port>/<db-name>?ssl=true&sslmode=require
# Database user which has the 'rds_iam' role
spring.datasource.username=<db user>
```
With this implementation, the `getPassword()` method will be called every 14 minutes.

## Conclusion

> Security is an important factor in many companies and environments and this is definitely a recommended way of securely accessing your database from inside your application.

Reviewer: ```suggestion
Security is an important factor in many companies and environments, and this is definitely a recommended way of securely accessing your database from inside your application.
```

> While it is slightly more of a hassle to set it up, it is definitely worth your while to implement RDS IAM authentication, especially if you are already running your application and infrastructure on AWS.

Reviewer: Maybe add that you don't have to worry about managing passwords anymore (e.g. rotating, ...)

---
layout: post
authors: [yolan_vloeberghs]
title: 'RDS IAM Authentication with Spring Boot - secure password-less database authentication on AWS'
image: /img/2022-06-13-aws-rds-iam-authentication-spring-boot.jpeg
tags: [Spring Boot, AWS, Spring, Security]
category: Cloud
comments: true
---

- [Introduction](#introduction)
- [Database role grant](#database-role-grant)
- [AWS](#aws)
- [Spring Boot](#spring-boot)
- [Conclusion](#conclusion)

## Introduction
Database authentication in Spring Boot has traditionally always been very simple to implement and configure to your needs.

> Whether it is with a custom data source, or just using the `spring.datasource` properties, it was always pretty straight forward to get a working database connection.

Reviewer 3: straightforward

Handling the sensitive data such as database passwords in different environments can be tricky, but nowadays, there are enough solutions to tackle this problem in a secure and discrete way.

> However, if you are deploying your application on the AWS cloud, you don't have to struggle passwords anymore or need a fancy tool, because you simply don't need a password.

Reviewer 3: However is volgens mij niet het juiste woord hier, zou eerder iets gebruiken als Fortunately of zo.

```json
{
   "Version": "2012-10-17",
   "Statement": [
      {
         "Effect": "Allow",
         "Action": [
             "rds-db:connect"
         ],
         "Resource": [
             "arn:aws:rds-db:region:account-id:dbuser:DbiResourceId/db-user-name"
         ]
      }
   ]
}
```

If you are running your application on an EC2 instance, make sure that your EC2 instance has an IAM role attached to it which explicitly gives access to the `rds-db:connect` action.
On EKS (Kubernetes for AWS), you can use [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html){:target="_blank" rel="noopener noreferrer"}.

## Spring Boot
To achieve the token fetching in Spring Boot, I created a custom `HikariDataSource` which overrides the `getPassword()` method and used the AWS Java SDK to create a request for an authentication token from RDS.

Note that for MariaDB, you do not need your own custom implementation as [they have an implementation of their own](https://mariadb.com/kb/en/mariadb-connector-j-250-release-notes/#aws-iam){:target="_blank" rel="noopener noreferrer"}. 

> 

Reviewer 4: Maybe you can mention what you need to add for Maven (pom.xml), which dependecy do you need to add for AWS Java SDK, and do you need all of it to work?
